library/src/scala/annotation/MainAnnotation.scala

```diff
@@ -0,0 +1,122 @@
+ *  It links a compiler-generated main method (call it compiler-main) to a user
+ *  written main method (user-main).
+ *  The protocol of calls from compiler-main is as follows:
+ *
+ *    - create a `command` with the command line arguments,
+ *    - for each parameter of user-main, a call to `command.argGetter`,
+ *      or `command.varargGetter` if is a final varargs parameter,
+ *    - a call to `command.run` with the closure of user-main applied to all arguments.
+ *
+ *  Example:
+ *  ```scala
+ *  /** Sum all the numbers
+ *   *
+ *   *  @param first Fist number to sum
+ *   *  @param rest The rest of the numbers to sum
+ *   */
+ *  @myMain def sum(first: Int, rest: Int*): Int = first + rest.sum
+ *  ```
+ *  generates
+ *  ```scala
+ *  object foo {
+ *    def main(args: Array[String]): Unit = {
+ *      val cmd = new myMain().command(
+ *        info = new CommandInfo(
+ *          name = "foo.main",
+ *          documentation = "Sum all the numbers",
+ *          parameters = Seq(
+ *            new ParameterInfo("first", "scala.Int", hasDefault=false, isVarargs=false, "Fist number to sum"),
+ *            new ParameterInfo("rest", "scala.Int" , hasDefault=false, isVarargs=true, "The rest of the numbers to sum")
+ *          )
+ *        )
+ *        args = args
+ *      )
+ *      val args0 = cmd.argGetter[Int](0, None) // using cmd.Parser[Int]
+ *      val args1 = cmd.varargGetter[Int] // using cmd.Parser[Int]
+ *      cmd.run(() => sum(args0(), args1()*))
+ *    }
+ *  }
+ *  ```
+ */
+@experimental
+trait MainAnnotation extends StaticAnnotation:
+  import MainAnnotation.*
+
+  /** The class used for argument string parsing and arguments into a `T` */
+  type Parser[T]
+
+  /** The required result type of the main method.
+   *
+   *  If this type is Any or Unit, any type will be accepted.
+   */
+  type Result
+
+  /** A new command with arguments from `args`
+   *
+   *  @param info The information about the command (name, documentation and info about parameters)
+   *  @param args The command line arguments
+   */
+  def command(info: CommandInfo, args: Array[String]): Command[Parser, Result]
+
+end MainAnnotation
+
+@experimental
+object MainAnnotation:
+
+  /** A class representing a command to run */
+  trait Command[Parser[_], Result]:
+
+    /** The getter for the `idx`th argument of type `T`
+     *
+     *   @param idx The index of the argument
+     *   @param defaultArgument Optional lambda to instantiate the default argument
+     */
+    def argGetter[T](idx: Int, defaultArgument: Option[() => T])(using Parser[T]): () => T
+
+    /** The getter for a final varargs argument of type `T*` */
+    def varargGetter[T](using Parser[T]): () => Seq[T]
+
+    /** Run `program` if all arguments are valid if all arguments are valid
+     *
+     *  @param program A function containing the call to the main method and instantiation of its arguments
+     */
+    def run(program: () => Result): Unit
```

Comment:
Not sure how to restrict the main method to return the `Result` type when `Result =:= Unit` as any type can be coerced into a `Unit`.

Maybe it should be 
```scala
def run[R](program: () => R)(using R <:< Result): Unit
```
But then we get error messages like
```
Cannot prove that Int <:< Unit.
```

Could also define something like this 
```scala
def run[R](program: () => R)(using ResultType[R]): Unit

@implicitNotFound("The result type of the main method must of type ${Result}")
class ResultType[T](using T <:< Result)
object ResultType:
  given [T](using T <:< Result): ResultType[T] = new ResultType
```
and fail with
```scala
  |The result type of the main method must of type Unit.
  |I found:
  |
  |    cmd.ResultType.given_ResultType_T[Int](/* missing */summon[Int <:< Unit])
  |
  |But no implicit values were found that match type Int <:< Unit.
```
Alternatively, it could be implemented as a macro and the type could be checked before generating the code. This seems a bit overkill.

Reply:
The current error message a `type Result = Int` is also not ideal.
```
  |Found:    () => Unit
  |Required: () => Int
```